File: LeetCode/496/main.cpp
#include <stack>
#include <unordered_map>

class Solution {
public:
  vector<int> nextGreaterElement(vector<int> &arr1, vector<int> &arr2) {
    vector<int> ans;
    stack<int> st;
    unordered_map<int, int> mp;
    int n = arr2.size();

    for (int i = n - 1; i >= 0; i--) {
      int ele = arr2[i];
      while (!st.empty() && st.top() < = ele) {
        st.pop();
      }
      int res = (st.empty()) ? -1 : st.top();

      // mp.insert({ele,res});
      mp[ele] = res;
      st.push(ele);
    }
    for (auto x : arr1) {
      ans.push_back(mp[x]);
    }
    return ans;
  }
};

File: LeetCode/496/main.py
from collections import deque, defaultdict


class Solution:
    # Use stack to keep track of elements that are smaller
    # than the one present on top of the stack
    def nextGreaterElement(self, arr1: List[int], arr2: List[int]) -> List[int]:
        n = len(arr2)
        st = deque()
        mp = defaultdict(int)
        ans = []

        st.append(arr2[n - 1])
        mp[arr2[n - 1]] = -1

        for i in range(n - 2, -1, -1):
            if arr2[i] < st[-1]:
                mp[arr2[i]] = st[-1]
                st.append(arr2[i])
            else:
                while len(st) > 0 and arr2[i] >= st[-1]:
                    st.pop()

                mp[arr2[i]] = st[-1] if len(st) > 0 else -1
                st.append(arr2[i])

        for x in arr1:
            ans.append(mp[x])

        return ans
